pulsar-broker/src/test/java/org/apache/pulsar/client/impl/KeySharedSubscriptionTest.java

```diff
@@ -0,0 +1,183 @@
+                .consumerName("con-1")
+                .messageListener((cons1, msg) -> {
+                    lastActiveTime.set(System.currentTimeMillis());
+                    nameToId.computeIfAbsent(cons1,(k) -> new ArrayList<>())
+                            .add(msg.getMessageId());
+                    recMessages.add(msg.getMessageId());
+                    if (canAcknowledgement.get()) {
+                        try {
+                            cons1.acknowledge(msg);
+                        } catch (PulsarClientException e) {
+                            throw new RuntimeException(e);
+                        }
+                    }
+                })
+                .subscribe();
+        @Cleanup
+        Consumer<byte[]> consumer2 = pulsarClient.newConsumer()
+                .topic(topic)
+                .subscriptionName("sub-1")
+                .subscriptionType(SubscriptionType.Key_Shared)
+                .messageListener((cons2, msg) -> {
+                    lastActiveTime.set(System.currentTimeMillis());
+                    nameToId.computeIfAbsent(cons2,(k) -> new ArrayList<>())
+                            .add(msg.getMessageId());
+                    recMessages.add(msg.getMessageId());
+                    if (canAcknowledgement.get()) {
+                        try {
+                            cons2.acknowledge(msg);
+                        } catch (PulsarClientException e) {
+                            throw new RuntimeException(e);
+                        }
+                    }
+                })
+                .consumerName("con-2")
+                .subscribe();
+        @Cleanup
+        Consumer<byte[]> consumer3 = pulsarClient.newConsumer()
+                .topic(topic)
+                .subscriptionName("sub-1")
+                .subscriptionType(SubscriptionType.Key_Shared)
+                .messageListener((cons3, msg) -> {
+                    lastActiveTime.set(System.currentTimeMillis());
+                    nameToId.computeIfAbsent(cons3,(k) -> new ArrayList<>())
+                            .add(msg.getMessageId());
+                    recMessages.add(msg.getMessageId());
+                    if (canAcknowledgement.get()) {
+                        try {
+                            cons3.acknowledge(msg);
+                        } catch (PulsarClientException e) {
+                            throw new RuntimeException(e);
+                        }
+                    }
+                })
+                .consumerName("con-3")
+                .subscribe();
+
+        @Cleanup
+        Producer<byte[]> producer = pulsarClient.newProducer()
+                .topic(topic)
+                .enableBatching(true)
+                .batchingMaxPublishDelay(5, TimeUnit.MILLISECONDS)
+                // We chose 999 because the maximum unacked message is 1000
+                .batchingMaxMessages(999)
+                .create();
+
+        for (int i = 0; i < totalMsg; i++) {
```

Comment:
Is it possible to reduce the number of `totalMsg`, this seems would increase unit test load significantly.

Reply:
fixed.